File: tests/snapshots/dataframe_basics.rs
// Native dataframes — read a CSV, compute new columns with Excel-style array
// formulas, filter rows, select, and pull a column out. Backed by polars (pure
// Rust — no Python). See dataframe_spec.md.

use vbr_stdlib::{DataFrame};

#[allow(unused_imports)]
use vbr_stdlib::dataframe::{col, lit, when};

fn main() {
    let mut df: DataFrame = DataFrame::read_csv("people.csv");
    let (rows, cols): (i64, i64) = df.shape();
    println!("loaded {} rows, {} columns", rows, cols);
    // Column formulas: arithmetic across whole columns, and an IIf band.
    df = df.with_column("total", col("price") * col("qty"));
    df = df.with_column("band", when(col("age").gt_eq(lit(18))).then(lit("adult")).otherwise(lit("minor")));
    // Row filter: a boolean mask over columns, combined with a Dim'd value.
    let cutoff: i64 = 30;
    df = df.filter(col("age").gt(lit(cutoff)).and(col("active")));
    df = df.select(&["name", "band", "total"]);
    df.print();
    let names: Vec<String> = df.column("name");
    println!("first kept: {}", names[0]);
    df.write_csv("out.csv");
    println!("wrote out.csv");
}

File: tests/snapshots/dataframe_basics.py
# Native dataframes — read a CSV, compute new columns with Excel-style array
# formulas, filter rows, select, and pull a column out. Backed by polars (pure
# Rust — no Python). See dataframe_spec.md.

from vbrpy import _vb, col, lit, read_csv, when

def main():
    df: object = read_csv('people.csv')
    rows, cols = df.shape
    print(f"loaded {_vb(rows)} rows, {_vb(cols)} columns")
    # Column formulas: arithmetic across whole columns, and an IIf band.
    df = df.with_columns((col("price") * col("qty")).alias('total'))
    df = df.with_columns(when((col("age") >= lit(18))).then(lit('adult')).otherwise(lit('minor')).alias('band'))
    # Row filter: a boolean mask over columns, combined with a Dim'd value.
    cutoff: int = 30
    df = df.filter(((col("age") > lit(cutoff)) & col("active")))
    df = df.select(['name', 'band', 'total'])
    print(df)
    names: list[str] = df['name'].to_list()
    print(f"first kept: {_vb(names[0])}")
    df.write_csv('out.csv')
    print('wrote out.csv')


if __name__ == "__main__":
    main()
